lib/mix/lib/mix/dep/lock.ex

```diff
@@ -32,48 +32,74 @@ defmodule Mix.Dep.Lock do
   """
   @spec read() :: map
   def read() do
-    case File.read(lockfile()) do
-      {:ok, info} ->
-        assert_no_merge_conflicts_in_lockfile(lockfile(), info)
+    lockfile = lockfile()
 
-        case Code.eval_string(info, [], file: lockfile()) do
-          {lock, _binding} when is_map(lock) -> lock
-          {_, _binding} -> %{}
-        end
+    case File.read(lockfile) do
+      {:ok, content} ->
+        read_lock(content, lockfile)
 
       {:error, _} ->
         %{}
     end
   end
 
+  @conflict_regex ~r/<{7} .+\n|\|{7} .+\n[\w\W\s]+={7}\n|={7}\n|>{7} .+\n/
+
+  defp read_lock(content, lockfile) do
+    if String.contains?(content, ~w(<<<<<<< ======= >>>>>>>)) do
+      resolve_merge_conflicts(content, lockfile)
+    else
+      eval_content(content, lockfile)
+    end
+  end
+
+  defp resolve_merge_conflicts(content, lockfile) do
+    new_map =
+      content
+      |> String.replace(@conflict_regex, "")
+      |> eval_content(lockfile)
+
+    write_lock(new_map, %{}, lockfile)
+
+    new_map
+  rescue
+    SyntaxError ->
```

Comment:
I think this slightly hides the flow. Instead of having `Code.eval_string/3` raise and rescue here, I'd go with having `eval_content/2` return `{:ok, new_map} | :error` and rescue in there:

```elixir
defp resolve_merge_conflicts(content, lockfile) do
  case content |> String.replace(@configt_regex, "") |> eval_content(lockfile) do
    {:ok, new _map} ->
      write_lock(new_map, %{}, lockfile)
      new_map
    :error ->
      Mix.raise(...)
  end
end

defp eval_content(content, lockfile) do
  case Code.eval_string(...) do
    ... -> {:ok, lock}
    ... -> {:ok, %{}}
  end
rescue
  SyntaxError -> :error
end
```